### PR TITLE
Refactor string DMA

### PR DESCRIPTION
**Pull request by gkorland**
1. Encapsulate the string when opened for Write  
2. Encapsulate string resize on write or append 

ref #248

### CODE REVIEW

File: src/key.rs [hunk 1]
```diff
@@ -75,13 +73,20 @@ impl RedisKey {
         self.key_inner == null_key
     }
 
-    pub fn read(&self) -> Result<Option<String>, RedisError> {
-        let val = if self.is_null() {
-            None
+    pub fn read<'a>(&self) -> Result<Option<&'a [u8]>, RedisError> {
```
Comment (MeirShpilraien) [inline src/key.rs:76]:
```suggestion
    pub fn read(&self) -> Result<Option<&'_ [u8]>, RedisError> {
```
Not sure whats the `'a` is

File: src/key.rs [hunk 2]
```diff
@@ -437,6 +437,56 @@ where
...
+        let dma = raw::string_dma(key.key_inner, &mut length, raw::KeyMode::WRITE);
+        if dma.is_null() {
+            Err(RedisError::Str("Could not read key"))
+        } else {
+            let buffer = unsafe { std::slice::from_raw_parts_mut(dma.cast::<u8>(), length) };
+            Ok(StringDMA { key, buffer })
+        }
+    }
+
+    pub fn write(&mut self, data: &[u8]) -> Result<&mut Self, RedisError> {
```
Comment (MeirShpilraien) [inline src/key.rs:457]: Why not implement `std::ops::index`and `std::ops::index_mut`?
https://doc.rust-lang.org/std/ops/trait.Index.html
https://doc.rust-lang.org/std/ops/trait.IndexMut.html

File: src/key.rs [hunk 3]
```diff
@@ -437,6 +437,56 @@ where
...
+            let dma = raw::string_dma(self.key.key_inner, &mut length, raw::KeyMode::WRITE);
+            self.buffer = unsafe { std::slice::from_raw_parts_mut(dma.cast::<u8>(), length) };
+        } else {
+            return Err(RedisError::Str("Failed to truncate string"));
+        }
+        self.buffer[current_len..new_len].copy_from_slice(data);
+        Ok(self)
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
```
Comment (MeirShpilraien) [inline src/key.rs:485]: We can implement Deref
https://doc.rust-lang.org/std/ops/trait.Deref.html

**Review by MeirShpilraien — COMMENTED** [review]
Small suggestions

File: src/key.rs (hunk 2 above)
Comment (gkorland) [reply src/key.rs:457]: It's an idea, but it can't be as efficient as a bulk write, I'll add it anyway.

File: src/key.rs (hunk 3 above)
Comment (gkorland) [reply src/key.rs:485]: do you need Deref? there's nothing to release here. StringDMA only holds borrows and its life time depends on the KeyHolder lifetime

Comment (MeirShpilraien) [reply src/key.rs:501]: If you implement deref, you do not need to implement is empt ly (for example)

Comment (gkorland) [reply src/key.rs:501]: is_empty is just for users simplicity

File: src/key.rs (hunk 2 above)
Comment (gkorland) [reply src/key.rs:473]: Any added Index and IndexMut too

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**